prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/repository/NodesClient.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.model.dataformat.JsonLibrary;
+import org.apache.hc.client5.http.HttpHostConnectException;
+import org.apache.hc.core5.http.MalformedChunkCodingException;
+import org.apache.hc.core5.http.NoHttpResponseException;
+import org.springframework.stereotype.Component;
+
+import org.alfresco.hxi_connector.common.exception.EndpointServerErrorException;
+import org.alfresco.hxi_connector.common.model.repository.NodeEntry;
+import org.alfresco.hxi_connector.common.util.ErrorUtils;
+import org.alfresco.hxi_connector.prediction_applier.config.NodesApiProperties;
+
+@Component
+@RequiredArgsConstructor
+public class NodesClient extends RouteBuilder
+{
+    public static final String DIRECT_ENDPOINT = "direct:" + NodesClient.class.getSimpleName();
+    private static final String RETRYABLE_ROUTE = "direct:retryable-" + NodesClient.class.getSimpleName();
+    static final String ROUTE_ID = "repository-nodes";
+    private static final String NODE_ID_HEADER = "nodeId";
+    private static final String URI_PATTERN = "%s/alfresco/api/-default-/public/alfresco/versions/1/nodes/${headers.%s}?httpMethod=PUT&authMethod=Basic&authUsername=%s&authPassword=%s&authenticationPreemptive=true&throwExceptionOnFailure=false";
+    private static final int EXPECTED_STATUS_CODE = 200;
+    public static final String UNEXPECTED_STATUS_CODE_MESSAGE = "Unexpected response status code - expecting: %d, received: %d";
+    private static final Set<Class<? extends Throwable>> RETRY_REASONS = Set.of(
+            EndpointServerErrorException.class,
+            UnknownHostException.class,
+            JsonEOFException.class,
+            MismatchedInputException.class,
+            HttpHostConnectException.class,
+            NoHttpResponseException.class,
+            MalformedChunkCodingException.class);
+
+    private final NodesApiProperties nodesApiProperties;
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void configure() throws Exception
+    {
+        // @formatter:off
+        onException(RETRY_REASONS.toArray(Class[]::new))
+            .retryAttemptedLogLevel(LoggingLevel.WARN)
+            .logExhaustedMessageBody(true)
+            .log(LoggingLevel.ERROR, log, "Unexpected response. Headers: ${headers}, Body: ${body}")
+            .maximumRedeliveries(nodesApiProperties.retry().attempts())
+            .redeliveryDelay(nodesApiProperties.retry().initialDelay())
+            .backOffMultiplier(nodesApiProperties.retry().delayMultiplier())
+            .stop();
+
+        onException(Exception.class)
+            .log(LoggingLevel.ERROR, log, "Unexpected response. Headers: ${headers}, Body: ${body}")
+            .stop();
+
+        from(DIRECT_ENDPOINT)
+            .setHeader(NODE_ID_HEADER, simple("${body.id}"))
+            .marshal()
+            .json(JsonLibrary.Jackson)
+            .to(RETRYABLE_ROUTE)
+            .end();
+
+        from(RETRYABLE_ROUTE)
```

Comment:
Is this route separated out (from the one starting on `DIRECT_ENDPOINT`) because of retry handling?

Reply:
Yes - Camel by default retries only the point where exception was thrown, not the whole route. Because in our case exception is thrown after call to endpoint:
```
.toD(URI_PATTERN.formatted(nodesApiProperties.baseUrl(), NODE_ID_HEADER, nodesApiProperties.username(), nodesApiProperties.password()))
.choice()
.when(header(HTTP_RESPONSE_CODE).isNotEqualTo(String.valueOf(EXPECTED_STATUS_CODE)))
    .process(this::throwExceptionOnUnexpectedStatusCode)
```
I had to extract this to separate sub-route and use the `.errorHandler(noErrorHandler())`